File: patientExamAggregator.py
from typing import Dict, List

#---------------Starting of Patient Class---------------#
class Patient:
    def __init__(self, id: str, name: str):
        self.id = id
        self.name = name
        self.exams: List[Exam] = []
#---------------Ending of Patient Class---------------#

#---------------Starting of Exam Class---------------#
class Exam:
    def __init__(self, id: str):
        self.id = id
#---------------Ending of Exam Class---------------#

#---------------Starting of PatientExamSystem Class---------------#
class PatientExamSystem:
    """
    Manages the Patient Record System.
    """
    def __init__(self):
        self.patients: Dict[str, Patient] = {}
        self.exams: Dict[str, Patient] = {}

    #-----Starting of add_patient() Method-----#
    def add_patient(self, id: str, name: str) -> None:
        """
        Function to add Patient Record to the System.
        Parameters:
            id(str): Patient ID
            name(str): Patient Name
        """
        if id not in self.patients:
            self.patients[id] = Patient(id, name)
    #-----Ending of add_patient() Method-----#

    #-----Starting of add_exam() Method-----#
    def add_exam(self, patient_id: str, exam_id: str) -> None:
        """
        Function to add Examination Record to the System.
        Parameters:
            patiend_id(str): Patient ID
        """
        if patient_id in self.patients and exam_id not in self.exams:
            self.patients[patient_id].exams.append(Exam(exam_id))
            self.exams[exam_id] = self.patients[patient_id]
    #-----Ending of add_exam() Method-----#

    #-----Starting of del_exam() Method-----#
    def del_patient(self, id: str) -> None:
        """
        Function to delete patient Record from the system.
        Parameters:
            id(str): Patient ID
        """
        if id in self.patients:
            for exam in self.patients[id].exams:
                self.exams.pop(exam.id, None)
            self.patients.pop(id, None)
    #-----Ending of del_exam() Method-----#

    #-----Starting of del_exam() Method-----#
    def del_exam(self, id: str) -> None:
        """
        Function to delete Examination record from the System.
        Parameters:
            id(str): Examination ID
        """
        if id in self.exams:
            self.exams[id].exams = [exam for exam in self.exams[id].exams if exam.id != id]
            self.exams.pop(id, None)
    #-----Starting of del_exam() Method-----#

#---------------Ending of PatientExamSystem Class---------------#

File: Java Code - Alternate Approach/PatientExamAggregator.java
import java.io.*;
import java.util.*;

public class PatientExamAggregator {

    // Map to store patient records
    private Map<Integer, String> patients = new LinkedHashMap<>();

    // Map to store exam records
    private Map<Integer, Integer> exams = new HashMap<>();

    // Function to add a new patient record
    public void addPatient(int id, String name) {
        if (!patients.containsKey(id)) {
            patients.put(id, name);
        }
    }

    // Function to add a new exam record
    public void addExam(int patientId, int examId) {
        if (patients.containsKey(patientId) && !exams.containsKey(examId)) {
            exams.put(examId, patientId);
        }
    }

    // Function to delete a patient record and all associated exam records
    public void deletePatient(int id) {
        patients.remove(id);
        exams.entrySet().removeIf(entry -> entry.getValue() == id);
    }

    // Function to delete an exam record
    public void deleteExam(int id) {
        exams.remove(id);
    }

    // Function to print a summary of all patients and their exam counts
    public void printPatientSummary() {
        for (Map.Entry<Integer, String> patient : patients.entrySet()) {
            int id = patient.getKey();
            String name = patient.getValue();
            int examCount = (int) exams.values().stream().filter(x -> x == id).count();
            System.out.println("Name: " + name + ", Id: " + id + ", Exam Count: " + examCount);
        }
    }

    // Function to process a single instruction from the input file
    public void processInstruction(String instruction) {
        String[] segments = instruction.split(" ");
        String command = segments[0];
        int id;
        String name;
        switch (command) {
            case "ADD":
                String recordType = segments[1];
                switch (recordType) {
                    case "PATIENT":
                        id = Integer.parseInt(segments[2]);
                        name = String.join(" ", Arrays.copyOfRange(segments, 3, segments.length));
                        addPatient(id, name);
                        break;
                    case "EXAM":
                        int patientId = Integer.parseInt(segments[2]);
                        int examId = Integer.parseInt(segments[3]);
                        addExam(patientId, examId);
                        break;
                }
                break;
            case "DEL":
                recordType = segments[1];
                switch (recordType) {
                    case "PATIENT":
                        id = Integer.parseInt(segments[2]);
                        deletePatient(id);
                        break;
                    case "EXAM":
                        int examId = Integer.parseInt(segments[2]);
                        deleteExam(examId);
                        break;
                }
                break;
        }
    }

    // Function to process the entire input file
    public void processInstructions(List<String> instructions) {
        for (String instruction : instructions) {
            processInstruction(instruction);
        }
    }


    //reads files and provides list of instructions
    public List<String> readInstructions(String filename) {
        List<String> instructions = new ArrayList<>();

        try {
            File inputFile = new File(filename);
            FileReader fr = new FileReader(inputFile);

            BufferedReader br = new BufferedReader(fr);
            String line;
            while( (line = br.readLine()) != null) {
                instructions.add(line);
            }

            fr.close();
        } catch (IOException e) {
            System.out.printf("exception occurred: " + e.getMessage());
        }

        return instructions;
    }

    // Sample test cases
    public static void main(String[] args) {
        PatientExamAggregator aggregator = new PatientExamAggregator();
        aggregator.processInstructions( aggregator.readInstructions("input5.txt"));
        aggregator.printPatientSummary();
    }
}